app.R

```diff
@@ -15,88 +15,35 @@ library(magrittr)
 # Load config file
 source('./UI_config.R')
 
+# Load UI Components
+source('./ui_components/selection_sidebar.R')
+
+# Load data models
+source('./models/car_data_frame.R')
+
 # Define UI for application that draws a histogram
 ui <- fluidPage(# Application title
-  titlePanel("Old Faithful Geyser Data"),
+  titlePanel("Census Data"),
+  
   
-  # Sidebar with a slider input for number of bins
   sidebarLayout(
-    sidebarPanel(
-      sliderInput(
-        "bins",
-        "Number of bins:",
-        min = 1,
-        max = 50,
-        value = 30
-      ),
-      # Insert geography
-      selectInput(
-        "input_state",
-        h3("Select state"),
-        choices = state_list,
-        selected = state_list[1]
-      ),
-      selectInput(
-        "input_county",
-        h3("Select county"),
-        choices = county_list,
-        selected = county_list[1]
-      ),
-      selectInput(
-        "input_year",
-        h3("Select data year"),
-        choices = year_list,
-        selected = year_list[1]
-      )
-    ),
+    # Sidebar with a slider input for number of bins, and drop down selectors
+    # for county state and year
+    selection_sidebar(state_list, county_list, year_list),
```

Comment:
The UI is still half-baked. Abstracting the UI out to a separate bit of code will make it harder to iterate. Cleaving the ui and server objects into separate files could make it more legible though.

Reply:
I have had the opposite experience. This works for the same reason a binary tree is fast to search, you can quickly eliminate most most of the choices. When its all in one place you have to consider the entire ui all at once. If there are 100 lines, you start at the top and have to understand what each line is doing until you reach the line you care about. If there are 5 function calls to 20 line functions, you can read the names of those 5 functions and decide one of them is relevant and discard the other two.

The best way to go fast is to take small steps. You wouldn't want to build the entire UI in one go, you're nearly guaranteed to get it wrong the first time. Instead, break it into small chunks you can't possibly mess up: a drop down selector, a slider, a check box, text entry. You make those small chunks, then compose them together. Then you realize this text box would really be better as a selector, so you change just that one component without affecting anything else.

When there are discrete components, you can consider each one in isolation, and separately how they are composed together. The iteration is easier when you can think about one small part at a time.

Reply:
But, taking small steps and assuming that it won't be right is what has already been done so far, and has worked well in prior Shiny apps I've built. Breaking things down to the smallest iterable pieces adds tree-traversal-and-management overhead, and the project is still at a place where the tiny pieces need to be considered jointly with the UI still.

Reply:
Found a good talk (20 mins) that gets to what I'm talking about. Its from an OO perspective, but the concepts map identically onto functional programming. https://brightonruby.com/2016/the-point-of-objects-john-cinnamond/